src/ray/common/asio/instrumented_io_context.h

```diff
@@ -0,0 +1,115 @@
+// Copyright 2017 The Ray Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <boost/asio.hpp>
+#include <limits>
+#include "absl/container/flat_hash_map.h"
+#include "absl/synchronization/mutex.h"
+#include "ray/common/ray_config.h"
+#include "ray/util/logging.h"
+
+/// Count, queueing, and execution statistics for an asio handler.
+struct HandlerStats {
+  // Counts.
+  int64_t cum_count = 0;
+  int64_t curr_count = 0;
+
+  // Execution stats.
+  int64_t cum_execution_time = 0;
+};
+
+/// Count and queueing statistics over all asio handlers.
+struct GlobalStats {
+  // Queue stats.
+  int64_t cum_queue_time = 0;
+  int64_t min_queue_time = std::numeric_limits<int64_t>::max();
+  int64_t max_queue_time = -1;
+};
+
+/// A mutex wrapper around a handler stats struct.
+struct GuardedHandlerStats {
+  // Stats for some handler.
+  HandlerStats stats;
+
+  // The mutex protecting the reading and writing of these stats.
+  // This mutex should be acquired with a reader lock before reading, and should be
+  // acquired with a writer lock before writing.
+  mutable absl::Mutex mutex;
+};
+
+/// A mutex wrapper around a handler stats struct.
+struct GuardedGlobalStats {
+  // Stats over all handlers.
+  GlobalStats stats;
+
+  // The mutex protecting the reading and writing of these stats.
+  // This mutex should be acquired with a reader lock before reading, and should be
+  // acquired with a writer lock before writing.
+  mutable absl::Mutex mutex;
+};
+
+/// A proxy for boost::asio::io_context that collects statistics about posted handlers.
+class instrumented_io_context : public boost::asio::io_context {
+ public:
+  /// A proxy post function that collects count, queueing, and execution statistics for
+  /// the given handler.
+  ///
+  /// \param handler The handler to be posted to the event loop.
+  /// \param name A human-readable name for the handler, to be used for viewing stats
+  /// for the provided handler. Defaults to UNKNOWN.
+  void post(std::function<void()> handler, const std::string &name = "UNKNOWN")
```

Comment:
I initially started to go that route, since there are several benefits to that approach, being:

1. Collection (most likely an array) of stats structs could be allocated at instantiation time, and concurrent access could proceed without a collection-level mutex.
2. Enums would result in faster lookups (indexing into preallocated and immutable array) and better compile-time type checking than an opaque string key.

Unfortunately, the handler identifiers associated with the [client](https://github.com/ray-project/ray/blob/ca3ac543fe9717c574f40774fdecb58d69e15189/src/ray/rpc/grpc_client.h#L29-L33) and [server](https://github.com/ray-project/ray/blob/ca3ac543fe9717c574f40774fdecb58d69e15189/src/ray/rpc/grpc_server.h#L30-L37) gRPC methods are built dynamically via the macros, and defining an enum that contains all of those client and server gRPC methods would essentially spread the method definitions across very disparate modules.

The closest that I came to making that tenable was an enum definition at the beginning of this header file, with the macros passing the corresponding enum value to be used as the handler identifier. Do you think I should give that a go?

Reply:
A string is easier to deal with than an enum, and it appears the performance is good enough here. So I think we should keep it as a string.

Reply:
That's my opinion as well.

Reply:
I guess it's because the bottleneck for the testing cases is not CPU for now. 
The same throughput doesn't mean they are using the same resources and given that we can have arbitrary functions, it can give different results under different workloads. The resource usage (mem/cpu) should also be included in the benchmark. Well, anyway, that's beyond the scope of this PR, so I'm fine with it as this now.

The string can introduce a typo kind of thing as well, maybe `const string&` in some central place to avoid literal string?

Reply:
>The same throughput doesn't mean they are using the same resources, and given that we can have arbitrary functions, it can give different results under different workloads. The resource usage (mem/CPU) should also be included in the benchmark.

I strongly agree there. After we've fixed the current high priority core and IO bugs, I'm hoping that we can prioritize a detailed audit of certain hot paths for the raylet and core worker since I'm sure there's a lot of low-hanging fruit. I wouldn't be surprised if we identified this proxy as adding unnecessary overhead, at which point we can work towards minimizing that identified overhead. Also, if you'd like to help me out with implementing some core microbenchmarks testing the performance of core components (not in Python), or improving the existing Python microbenchmarks, I'd be happy to partner on that!

>The string can introduce a typo kind of thing as well, maybe const string& in some central place to avoid literal string?

I don't think that I follow, what would defining all string literals in a central place give us? Couldn't those literal string typos happen just as easily in that central place?

Reply:
> I don't think that I follow, what would defining all string literals in a central place give us? Couldn't those literal string typos happen just as easily in that central place?

I originally thought that there might be a pattern like this:

```
// some place
post("metricsA")
// another place
post("metricsA")
```
With this pattern, it'll avoid typo kind of thing if we do it in this way

```
metricsA="metricsA"

// some place 
post(metricsA)
// another place
post(metricsA)
```
But it looks like it's not the case here. All posts are measuring different metrics.

Reply:
Ah got it, yeah each string literal should be defined once, at the point of posting.